pbj-core/pbj-grpc-helidon/docs/design.md

```diff
@@ -0,0 +1,450 @@
+            response.onError(throwable);
+        }
+
+        @Override
+        public void onComplete() {
+            response.onNext(HelloReply.newBuilder()
+                    .setMessage("Hello " + String.join(", ", names))
+                    .build());
+            response.onComplete();
+        }
+    };
+}
+```
+
+When the client streaming handler is built, it immediately calls the application method for that RPC call. That method
+return a `Flow.Subscriber`, F1. Then, as each message is received from the client, the `onNext` method of the request
+convert is called, which then calls `onNext` on F1 after converting from `Bytes` to the request object. When F1 calls
+any method on the `response` subscriber, the response is converted back to `Bytes` and sent back to the client.
+
+![Client Streaming Pipeline](images/client-streaming.png)
+
+## Bidirectional Streaming Pipeline
+
+Bidirectional stream methods have the same signature as the client streaming methods, but they permit multiple
+messages to be returned to the client. The rate at which messages are received from the client and sent to the client
+can be arbitrary. For example, one thread may push messages to the client while the request message is receiving
+messages from the client. Or each message from the client may trigger a response message to be sent to the client.
+How this is handled is up to the application developer.
+
+For example, the following is a bidirectional streaming method, where for every request, there is a response, but
+using the same thread. Here you can see that in the method call (`sayHelloStreamBidi`), we receive a `Flow.Subscriber`.
+Using the `onNext`, `onError`, and `onComplete` methods of this subscriber we can send messages back to the client,
+terminate the connection due to errors, or close the connection gracefully. We also return a new `Flow.Subscriber` that
+will receive messages from the client. The PBJ-Helidon module will push messages from the client to this subscriber
+using the `onNext` method.
+
+```java
+public class GreeterServiceImpl implements GreeterService {
+    @Override
+    @NonNull
+    public Flow.Subscriber<HelloReply> sayHelloStreamBidi(@NonNull final Flow.Subscriber<HelloRequest> request) {
+        return new Flow.Subscriber<>() {
+            @Override
+            public void onSubscribe(Flow.Subscription subscription) {
+                subscription.request(Long.MAX_VALUE); // turn off flow control if you want to
+            }
+
+            @Override
+            public void onNext(HelloRequest item) {
+                replies.onNext(HelloReply.newBuilder()
+                        .setMessage("Hello " + item.getName())
+                        .build());
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                replies.onError(throwable);
+            }
+
+            @Override
+            public void onComplete() {
+                replies.onComplete();
```

Comment:
`s/replies/request/` at lines 361, 368, and 373? Or where does the `replies` come from otherwise?

Reply:
Good o'l CoPilot, innovating with variable names :). Will fix.